### DISCUSSION TITLE
[RFC] deep watch allow setting condition for skipping object

### DISCUSSION THREAD

**Discussion by kimverchan** [vuejs/rfcs#712] [RFC Discussions]
Start Date: 2024-10-15
Target Major Version: 3.x
Implementation PR: https://github.com/vuejs/core/pull/12157

## Summary

If you want to skip the dependency collection of an object when deeply watch to an object, you cannot do so currently. Add the skip attribute to watchOptions to do this. 

## Examples

``` javascript
let dummy = 0
const source = reactive({
    num1: 1,
    nested: { _skip: true, num2: 0 },
})
watch(source, () => dummy++, {
    deep: true,
    skip(val) {
      return val._skip // return true to skip val
    }
})
source.nested.num2++ // do not run callback
expect(dummy).toBe(0)
source.num1++ // run callback
expect(dummy).toBe(1)
source.nested.num2++  // do not run callback
expect(dummy).toBe(1)
```

[Result Playground](https://deploy-preview-12157--vue-sfc-playground.netlify.app/#eNp9kk1v2zAMhv8KoUsdxLAb92Y4wT7Qw3bYhm1HAYMn04kbWzL04WYw/N9HSUnaFWt9ksjnJV+amtn7ccwmh6xklRG6Gy0YtG7ccdkNo9IWZtDYpvBYW3FI6VwL200IC7RaDXBD2hsuuRRKGguNG4Y/sPWS5Hb1FDfKaYEhEfXJzCXQJ92wKWGTnm9oLDYl9fxljt1YgtUOUw8VJdzCQthCVYOXJNZMIVnBdhc7Z1PdO1yvUziXbxAvVWLAl02IWl0IIEvWaQkUzEJXyPNLzAvBqqDyQJQs0QWXrZM0jJLQyg25CBWjq8zPtV4T+IIrXnJhYo8XZ7zK4x5oA3SxOIx9bZFuANVvZy2VeSf6Thy3nFFfznbiUMs9bqo8pl9Diyta/Is23bSb5/PulqXKfcAbuTZnKbOGNtl2++zBKEmvJQzBmVDD2PWov45+QMMZLS/+Jc7qvlePn0Ps2QZIc0Bx/E/8wZx8jLNvGg3qCTm75myt92hj+v7HFzzR+ZocVON6ot9Ifkejeuc9RuyDkw3ZfsYFt5/Cm+/k/qe5P1mU5jKUN+pJ/wQ9Tc/+4xujP9m9y+6CjnbLlr+8LyTr)

## Motivation

When using the `watch` function to observe objects in depth, suppose that the object to be observed contains a specific object that does not need to be observed, but the specific object cannot be determined by depth. In this case, you can use the `skip` option to control the monitoring conditions to avoid unnecessary observations.

**Comment by posva** [comment]
The motivation part should include practical real world examples. Right now it repeats what can be read with the simple examples.
